### PR TITLE
feat: Support connection sharing across gRPC channels

### PR DESCRIPTION
**Pull request by amanda-tarafa**
@jskeet the first commit is some helpful refactoring.

For the second commit, naming ideas are welcomed.

### CODE REVIEW

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClientBuilder.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using static Google.Cloud.PubSub.V1.ClientBuilderChannelHelper;
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClientBuilder.cs:21]: I think it *might* be clearer to just qualify the calls to the methods. It's probably clear in VS, but I was a bit confused to start with.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientBuilder.cs [hunk 2]
```diff
@@ -65,6 +56,12 @@ public int? ClientCount
         }
     }
 
+    /// <summary>
+    /// Whether the <see cref="PublisherServiceApiClient"/>s may share gRPC connections or not.
+    /// Defaults to false, meaning, each client has its own gRPC connection.
+    /// </summary>
+    public bool ClientsShareConnections { get; set; }
```
Comment (jskeet) [inline apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientBuilder.cs:63]: Consider ShareClientConnections instead? I'm not massively concerned about either option though.

**Review by jskeet — APPROVED** [review]
Please change the message for the second commit before merging as that'll end up in release notes.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientBuilder.cs (hunk 2 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClientBuilder.cs:63]: Yes, I'll change to that.

**Comment by amanda-tarafa** [comment]
Latest changes done:

- Property name changed from ClientsShareConnections => ShareClientConnections
- Commit message and PR title changed as well

Merging now, thanks!

**Pull request merged**